app/controllers/CompilerService.scala

```diff
@@ -203,14 +205,23 @@ object CompilerService extends Controller {
   private def extractModelString(argMap: Map[String, String], missingMsg: String): ValidationNel[String, String] =
     (argMap get modelKey).fold(missingMsg.failureNel[String])(_.successNel[String])
 
-  private def fetchURL(url: String): ValidationNel[String, String] = {
-    Try(new URL(url)) map {
-      wellFormedURL => usingSource(_.fromURL(wellFormedURL))(_.mkString).successNel
-    } recover {
-      case _: MalformedURLException => s"'$url' is an invalid URL.".failureNel
-    } getOrElse {
-      s"An unknown error occurred while processing the URL '$url'. Make sure the url is publicly accessible.".failureNel
-    }
+  private def fetchURL(url: String, hostUri: String): ValidationNel[String, String] = {
+    val locallyHostedRegex = s"^https?://$hostUri([A-Za-z0-9%/]+\\.nlogo)$$".r
+    locallyHostedRegex.findFirstIn(url).flatMap {
+      case locallyHostedRegex(file) =>
+        Assets.resourceNameAt("public", file).map(path =>
+          usingSource(_.fromFile(path))(_.mkString).successNel
```

Comment:
Yeah. I was continuing to work on that locally, but the options seemed to be using package-private methods (which is gross) or bubbling an Action up the call stack (which seemed even worse). I'll give it another try using `Play.resource`, which I tried briefly, but didn't seem to work as I had expected.